### PR TITLE
Add jssets

### PR DESCRIPTION
**Pull request by juancarlospaco**
- `Set` for the JavaScript target https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

See https://github.com/nim-lang/Nim/pull/16339#issue-538769691 and  https://github.com/nim-lang/Nim/pull/16339#issuecomment-744485709 and https://github.com/nim-lang/fusion/pull/57#issue-539766026 and https://github.com/nim-lang/fusion/pull/56#issue-539747803 and https://github.com/nim-lang/fusion/pull/55#issue-539726540
and https://github.com/nim-lang/fusion/pull/58#issue-539886249

### CODE REVIEW

File: src/fusion/js/jssets.nim [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+## * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
+when not defined(js) and not defined(nimdoc):
+  {.fatal: "Module jssets is designed to be used with the JavaScript backend.".}
+
+type JsSet* {.importjs: "Set".} = ref object of JsRoot ## Set API.
+  size: cint
+
+func newJsSet*(): JsSet {.importjs: "new Set()".} ## Constructor for `JsSet`.
+
+func newJsSet*(items: openArray[SomeNumber]): JsSet {.importjs: "new Set(#)".} ## Constructor for `JsSet`.
```
Comment (timotheecour) [inline src/fusion/js/jssets.nim:11]: why SomeNumber? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set allows other types, including cstring

> The Set object lets you store unique values of any type, whether primitive values or object references.

Comment (juancarlospaco) [reply src/fusion/js/jssets.nim:11]: Need to think how the API from the Nim side should be, I see 2 ways:

- Only wrap for numeric types, code is simple and tiny.
- Allow any arbitrary types like JS, code is complicated.

:thinking:

Comment (timotheecour) [reply src/fusion/js/jssets.nim:11]:
```nim
import jsffi
func newJsSet*(items: openArray[JsObject]): JsSet {.importjs: "new Set(#)".} ## Constructor for `JsSet`.

func toString*(this: JsSet): cstring {.importjs: "Array.from(#).join(', ')".}
  # maybe improve this ?

runnableExamples:
  # ...
  let a: JsSet = newJsSet([1.toJs, 2.toJs, "asdf".toJs])
  echo a.toString()
```

IMO this is actually simpler; it just treats everything in the set as a JsObject, thus no need for getInt, getFloat etc

File: src/fusion/js/jssets.nim [hunk 2]
```diff
@@ -3,59 +3,45 @@
...
+    doAssert b.size == 3
+    a.toString() == @["1".cstring, "2", "3", "4"]
+    b.toString() == @["1.0".cstring, "2.0", "3.0"]
+    a.clear()
+    b.clear()
+    let d: JsSet = newJsSet([1.toJs, 2.toJs, 3.toJs])
+    doAssert d.size == 3
+    d.add(4.toJs)
+    d.delete(2.toJs)
+    doAssert 3.toJs in d
```
Comment (timotheecour) [inline src/fusion/js/jssets.nim:46]: `doAssert "3".cstring.toJs notin d`
(adds a negative example and illustrates that "3".cstring.toJs wouldn't match even though 3.toJs would match)

File: src/fusion/js/jssets.nim [hunk 3]
```diff
@@ -3,59 +3,45 @@
...
-
-    block:
-      let a = newJsSet([1, 2])
-      a.add(3.5)
-      doAssert a.getInt() == @[1, 2, 3]
-
-      let b = newJsSet([1.0, 2.0])
-      b.add(3)
-      doAssert b.getFloat() == @[1.0, 2.0, 3.0]
+    let a: JsSet = newJsSet([1.toJs, 2.toJs, 3.toJs, 4.toJs])
```
Comment (timotheecour) [inline src/fusion/js/jssets.nim:34]: lots of errors when i run
nim doc -b:js src/fusion/js/jssets.nim
or
nim doc -b:js --doccmd:-d:fusionJsSetTests src/fusion/js/jssets.nim

`proc size()` should be exposed (or field size made public but exposing a proc might make more sense to avoid setting size unless that has some meaning)


a.toString() == @["1".cstring, "2", "3", "4"] has no doAssert

import jsffi is missing from runnableExamples

etc...

basically we really need to start running CI and nim doc for js in fusion; it shouldn't be hard; a dedicated PR that does just that is welcome

File: src/fusion/js/jssets.nim [hunk 4]
```diff
@@ -3,59 +3,45 @@
...
   ## https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/has
 
-func getInt*(this: JsSet): seq[int] {.importjs: """
-  (() => {const x = []; #.forEach(a => x.push(parseInt(a))); return x})()""".}
-  ## Convert `JsSet` to `seq[int]`, all items will be converted to `int`.
-
-func getFloat*(this: JsSet): seq[float] {.importjs: """
-  (() => {const x = []; #.forEach(a => x.push(parseFloat(a))); return x})()""".}
-  ## Convert `JsSet` to `seq[float]`, all items will be converted to `float`.
+func toString*(this: JsSet): seq[cstring] {.importjs: """
```
Comment (timotheecour) [inline src/fusion/js/jssets.nim:27]: making `toString` return a seq[cstring] is too surprising. Either remove this or (better) implement `$(this: JsSet): string` (string intended, not cstring), that would behave similar to stdlib

File: src/fusion/js/jssets.nim (hunk 3 above)
Comment (timotheecour) [reply src/fusion/js/jssets.nim:36]: this problem is still there

nim doc -b:js src/fusion/js/jssets.nim
Error: undeclared field: 'toJs' for type system.int

**Comment by timotheecour** [comment]
@juancarlospaco you're re-requesting a review but https://github.com/nim-lang/fusion/pull/59#discussion_r549027926 is still un-addressed
```
nim doc -b:js --doccmd:-d:fusionJsSetTests src/fusion/js/jssets.nim
```
still fails, and my review comment says how you can easily fix it

File: src/fusion/js/jssets.nim [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    a.clear()
+    b.clear()
+    let d: JsSet = newJsSet([1.toJs, 2.toJs, 3.toJs])
+    doAssert d.size == 3
+    d.add(4.toJs)
+    d.delete(2.toJs)
+    doAssert 3.toJs in d
+    doAssert "3".cstring.toJs notin d
+    doAssert d.contains 1.toJs
+    echo $d  # @["1", "3", "4"]
```
Comment (timotheecour) [inline src/fusion/js/jssets.nim:52]: instead, `doAssert $d == ...`

File: src/fusion/js/jssets.nim [hunk 6]
```diff
@@ -0,0 +1,52 @@
+## `Set` for the JavaScript target.
+## * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
+when not defined(js) and not defined(nimdoc):
+  {.fatal: "Module jssets is designed to be used with the JavaScript backend.".}
+
+import jsffi
+
+type JsSet* {.importjs: "Set".} = ref object of JsRoot ## Set API.
+  size*: cint
```
Comment (timotheecour) [inline src/fusion/js/jssets.nim:9]: setting `size` should be disallowed, instead, add a proc
```nim
func len*(a: JsSet): int =
  a.size
```

**Review by timotheecour — APPROVED** [review]
LGTM; we really need to start running CI and nim doc for js in fusion otherwise all those tests are left un-tested

File: changelog.md [hunk 7]
```diff
@@ -9,4 +9,6 @@
   was previously only available in the `nimscript` module.
 
 
+- Add `jssets` module, Set for the JavaScript target
```
Comment (xflywind) [inline changelog.md:12]:
```suggestion
- Added `jssets` module, Set for the JavaScript target
```

File: src/fusion/js/jssets.nim [hunk 8]
```diff
@@ -0,0 +1,54 @@
+## `Set` for the JavaScript target.
+## * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
+when not defined(js) and not defined(nimdoc):
+  {.fatal: "Module jssets is designed to be used with the JavaScript backend.".}
+
+import jsffi
```
Comment (xflywind) [inline src/fusion/js/jssets.nim:6]: nit: std/jsffi

File: src/fusion/js/jssets.nim [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  (() => {const result = []; #.forEach(item => result.push(JSON.stringify(item))); return result})()""".}
+  ## Convert `JsSet` to `seq[cstring]`, all items will be converted to `cstring`.
+
+func len*(this: JsSet): int = this.size.int
+
+func `$`*(this: JsSet): string = $this.toString()
+
+
+runnableExamples:
+  import jsffi
```
Comment (xflywind) [inline src/fusion/js/jssets.nim:37]: ditto

**Review by xflywind — APPROVED** [review]

**Comment by juancarlospaco** [comment]
@timotheecour Ping. Also https://github.com/nim-lang/fusion/pull/55#pullrequestreview-560565775

**Review by timotheecour — APPROVED** [review]

File: src/fusion/js/jssets.nim (hunk 4 above)
Comment (timotheecour) [reply src/fusion/js/jssets.nim:27]: @juancarlospaco this was marked as resolved but the problem is still there.

Same remark as @araq's comment here https://github.com/nim-lang/Nim/pull/16409#discussion_r549346583 except this is worse because it returns a `seq[cstring]`
